src/merkle/tiered_smt/mod.rs

```diff
@@ -0,0 +1,396 @@
+use super::{
+    BTreeMap, BTreeSet, EmptySubtreeRoots, Felt, MerkleError, MerklePath, NodeIndex, Rpo256,
+    RpoDigest, StarkField, Vec, Word, EMPTY_WORD, ZERO,
+};
+use core::cmp;
+
+#[cfg(test)]
+mod tests;
+
+// TIERED SPARSE MERKLE TREE
+// ================================================================================================
+
+/// Tiered (compacted) Sparse Merkle tree mapping 4-element keys to 4-element values.
+///
+/// Leaves in the tree can exist only on specific depths called "tiers". These depths are: 16, 32,
+/// 48, and 64. If multiple values have keys with a common 64-bit prefix, such key-value pairs are
+/// stored in a sorted list at the last tier (depth = 64).
+///
+/// All initial values in the Tiered SMT are assumed to be [ZERO; 4], and thus, the root of of an
+/// empty tree will be the same as a root of a fully materialized Merkle tree with depth 64 where
+/// each leaf is set to [ZERO; 4].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TieredSmt {
+    root: RpoDigest,
+    nodes: BTreeMap<NodeIndex, RpoDigest>,
+    upper_leaves: BTreeMap<NodeIndex, RpoDigest>, // node_index |-> key map
+    bottom_leaves: BTreeMap<u64, BottomLeaf>,     // leaves of depth 64
+    values: BTreeMap<RpoDigest, Word>,
+}
+
+impl TieredSmt {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    /// The number of levels between tiers.
+    const TIER_SIZE: u8 = 16;
+
+    /// Depths at which leaves can exist in a tiered SMT.
+    const TIER_DEPTHS: [u8; 4] = [16, 32, 48, 64];
+
+    /// Maximum node depth. This is also the bottom tier of the tree.
+    const MAX_DEPTH: u8 = 64;
```

Comment:
I think we had discussed potentially leveraging the `TieredSmt` for use cases in which we would want the tiers placed at different levels.  How hard would it be to make this solution generic over the tier configuration using const generics?

Reply:
Shouldn't be too difficult. There are a few methods which do assume this specific structure - but I think they can be relatively easily updated. I'd leave this to subsequent PRs though.